### ISSUE TITLE
Calling unsubscribe() in transportFailure handler will not close all connections and client keeps trying to reconnect

### ISSUE THREAD

**Issue by ookke** [Atmosphere/atmosphere-javascript#153]
It seems that calling unsubscribe() in a transportFailure handler will not cancel all connections, atmosphere will go ahead and try to reconnect anyway. I would expect atmosphere to give up all attempts when I call unsubscribe().

I can reproduce this with a offline app that is loaded from the appcache and the server not accepting any connections. I guess something goes wrong in the unsubscribe cleanup in the situation that the current connection attempt is not successfully established yet?

Console output from the client:
```js
 WebSocket connection to 'wss://localhost/app/api/v1/notifications/?X-Atmosphere-tracking-id=0…e-TrackMessageSize=true&Content-Type=application/json&X-atmo-protocol=true' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
app.js:68429 Fri Feb 20 2015 14:40:37 GMT+0200 (FLE Standard Time) Atmosphere: websocket.onerror
app.js:68429 Fri Feb 20 2015 14:40:37 GMT+0200 (FLE Standard Time) Atmosphere: websocket.onclose
app.js:68429 Websocket closed, reason: Connection was closed abnormally (that is, with no close frame being sent).
app.js:68429 Websocket closed, wasClean: false
app.js:68429 Fri Feb 20 2015 14:40:37 GMT+0200 (FLE Standard Time) Atmosphere: Firing onClose
app.js:68429 Fri Feb 20 2015 14:40:37 GMT+0200 (FLE Standard Time) Atmosphere: Closed but not firing onClose
app.js:68429 Websocket failed. Downgrading to Comet and resending

//transport failure fires here and we call socket.unsubscribe(), however atmosphere keeps going inifinitely with reconnects

app.js:68429 Closing
app.js:68429 Fri Feb 20 2015 14:41:23 GMT+0200 (FLE Standard Time) Atmosphere: Closed but not firing onClose
app.js:68429 Fri Feb 20 2015 14:41:23 GMT+0200 (FLE Standard Time) Atmosphere: Closed but not firing onClose
app.js:67299 GET https://localhost/app/api/v1/notifications/?X-Atmosphere-tracking-id…=true&Content-Type=application%2Fjson&X-atmo-protocol=true&_=1424436089000 net::ERR_CONNECTION_REFUSED
app.js:68429 Fri Feb 20 2015 14:41:29 GMT+0200 (FLE Standard Time) Atmosphere: ajaxRequest.onreadystatechange, new state: 4
app.js:68429 Fri Feb 20 2015 14:41:29 GMT+0200 (FLE Standard Time) Atmosphere: ajaxrequest.onerror
app.js:68429 Fri Feb 20 2015 14:41:29 GMT+0200 (FLE Standard Time) Atmosphere: Firing onReconnect
app.js:68429 Fri Feb 20 2015 14:41:29 GMT+0200 (FLE Standard Time) Atmosphere: Firing onReconnect
app.js:34616 POST https://localhost/app/api/v1/log net::ERR_CONNECTION_REFUSED
app.js:67299 GET https://localhost/app/api/v1/notifications/?X-Atmosphere-tracking-id…=true&Content-Type=application%2Fjson&X-atmo-protocol=true&_=1424436090029 net::ERR_CONNECTION_REFUSED
app.js:68429 Fri Feb 20 2015 14:41:30 GMT+0200 (FLE Standard Time) Atmosphere: ajaxRequest.onreadystatechange, new state: 4
app.js:68429 Fri Feb 20 2015 14:41:30 GMT+0200 (FLE Standard Time) Atmosphere: ajaxrequest.onerror
app.js:68429 Fri Feb 20 2015 14:41:30 GMT+0200 (FLE Standard Time) Atmosphere: Firing onReconnect
app.js:68429 Fri Feb 20 2015 14:41:30 GMT+0200 (FLE Standard Time) Atmosphere: Firing onReconnect
app.js:67299 GET https://localhost/app/api/v1/notifications/?X-Atmosphere-tracking-id…=true&Content-Type=application%2Fjson&X-atmo-protocol=true&_=1424436091114 net::ERR_CONNECTION_REFUSED
app.js:68429 Fri Feb 20 2015 14:41:32 GMT+0200 (FLE Standard Time) Atmosphere: ajaxRequest.onreadystatechange, new state: 4
app.js:68429 Fri Feb 20 2015 14:41:32 GMT+0200 (FLE Standard Time) Atmosphere: ajaxrequest.onerror
app.js:68429 Fri Feb 20 2015 14:41:32 GMT+0200 (FLE Standard Time) Atmosphere: Firing onReconnect
```

**Comment by ookke** [comment]
Actually https://github.com/Atmosphere/atmosphere-javascript/issues/150 is a duplicate of this issue, but the problem seems to be more severe than I originally reported. It's very hard to write own recovery logic for atmosphere connections when you can't rely on unsubscribe() cleaning up properly.

**Comment by jfarcand** [comment]
@ookke Why would you  `unsubscribe` if the transport fail? You should just switch transport, no?

**Comment by ookke** [comment]
It's a workaround for (https://github.com/Atmosphere/atmosphere-javascript/issues/151), I'm basically re-subscribing atmosphere with websockets again if it seems that the client supports websockets. Can I  just set the transport back to websocket in the transport failure handler?

**Comment by jfarcand** [comment]
@ookke try it :-)

**Comment by jfarcand** [comment]
@ookke It should work but make sure you don't reconnect forever trying to reuse the same tranbsport over and over

**Comment by ookke** [comment]
It doesn't seem to work very well, it will retry the websocket once and give up it seems.

**Comment by ookke** [comment]
Looks like websocket.onclose() gets called with _response.state = 'closed' which hits the condition at https://github.com/Atmosphere/atmosphere-javascript/blob/master/modules/javascript/src/main/webapp/javascript/atmosphere.js#L1470

**Comment by jfarcand** [comment]
@ookke I'm not sure I understand what you are trying to do. Can you share a test case on Github?

**Comment by ookke** [comment]
Well basically i'm trying to find a working workaround for https://github.com/Atmosphere/atmosphere-javascript/issues/151, I don't want websocket clients to downgrade to long polling due to network issues. 

I'm starting to think that it's perhaps not worth the effort and I should just live with the clients downgrading for now. I'll see if I have time to create a self-contained sample.

I still think unsubscribe() should always do a proper cleanup of all connections regardless where it's called from in case someone is trying to do strange things, it would make the framework more flexible.

**Comment by jfarcand** [comment]
@ookke Can you elaborate on your last paragraph? BTW just set the transport == `websocket` in yoor `onTransportFailure` if you don't want any fallback.
```js
    request.onTransportFailure = function(errorMsg, request) {
        request.fallbackTransport = "websocket";
```

**Comment by ookke** [comment]
What I was trying to say was that the Javascript api user would probably expect calling unsubscribe() to be guaranteed to stop all Atmosphere connections and connection attempts which wasn't the case when I reported the issue. Atmosphere would in some cases keep trying to reconnect anyway even after unsubscribe() has been called.

Anyway this issue is not relevant for me anymore, we switched to another library since we couldn't get the Atmosphere reconnect behavior to match our needs.
